src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -246,27 +287,22 @@ impl CopyToS3Uploader {
         let buffer_length = u64::cast_from(self.buf.len());
 
         if self.current_file_uploader.is_none() {
-            self.start_new_file().await?;
+            self.start_new_file_upload().await?;
         }
-        // Ideally it would be nice to get a `&mut uploader` returned from the `start_new_file`,
-        // but that runs into borrow checker issues when trying to add the `&self.buf` to the
-        // `uploader.add_chunk`.
-        let Some(uploader) = self.current_file_uploader.as_mut() else {
+
+        let Some(file_uploader) = self.current_file_uploader.as_ref() else {
             unreachable!("uploader initialized above");
         };
-        if buffer_length <= uploader.remaining_bytes_limit() || uploader.added_bytes() == 0 {
-            // Add to ongoing upload of the current file if still within limit.
-            // Or in the unlikely even the size of a single row is more than the max_file_size
-            // upload it anyway.
-            uploader.buffer_chunk(&self.buf).await?;
+        if file_uploader.can_add_more_bytes(buffer_length) || file_uploader.added_bytes() == 0 {
+            // Add to ongoing upload of the current file if still within max_file_size limit.
+            // Or in the unlikely event the size of a single row is more than the max_file_size
+            // and no data has been added to the uploader yet, upload it anyway.
+            self.upload_buffer().await?;
```

Comment:
instead of `can_add_more_bytes()` being public, can return an enum error type from `upload_buffer` if its too big (or 0)? then we can do:
```
 match self.upload_buffer().await {
    Err(TooManyBytes) => {
  		self.start_new_file_upload().await?;
		// If the buffer contains too many bytes for a fully new file, then there is some fundamental misconfiguration error, 
		// and we error permanently.
    	self.upload_buffer().await?;
	}
	...
}
```

Reply:
The thing is, for certain cases we are fine exceeding the max_file_size, so it's not always an error scenario. Or maybe we can pass a flag explicitly to `upload_buffer` in those cases to ignore max_file_size limits. Let me rejig this.

Reply:
Refactored to return `UploadExceedsMaxFileLimit` error variant and retrying with new file upon that.

Reply:
It's looking much simpler!